File: robocode-tankroyale-bot-api-java/src/main/java/dev/robocode/tankroyale/botapi/BaseBot.java
package dev.robocode.tankroyale.botapi;

import dev.robocode.tankroyale.botapi.events.BotEvent;
import dev.robocode.tankroyale.botapi.events.Condition;
import dev.robocode.tankroyale.botapi.internal.BaseBotInternals;

import java.net.URI;
import java.util.Collection;

/**
 * Abstract bot class that takes care of communication between the bot and the server, and sends
 * notifications through the event handlers. Most bots can inherit from this class to get access to
 * basic methods.
 */
public abstract class BaseBot implements IBaseBot {

  final BaseBotInternals __baseBotInternals;

  /**
   * Constructor for initializing a new instance of the BaseBot class, which should be used when
   * both BotInfo and server URL is provided through environment variables, i.e., when starting up
   * the bot using a bootstrap. These environment variables must be set to provide the server URL
   * and bot information, and are automatically set by the bootstrap tool for Robocode.
   *
   * <p><b>Example of how to set the predefined environment variables:</b>
   *
   * <p>ROBOCODE_SERVER_URL=ws://localhost<br>
   * BOT_NAME=MyBot<br>
   * BOT_VERSION=1.0<br>
   * BOT_AUTHOR=fnl<br>
   * BOT_DESCRIPTION=Sample bot<br>
   * BOT_URL=https://mybot.somewhere.net<br>
   * BOT_COUNTRY_CODE=DK<br>
   * BOT_GAME_TYPES=melee,1v1<br>
   * BOT_PLATFORM=Java<br>
   * BOT_PROG_LANG=Java 8<br>
   */
  public BaseBot() {
    __baseBotInternals = new BaseBotInternals(this, null, null);
  }

  /**
   * Constructor for initializing a new instance of the BaseBot class, which should be used when
   * server URL is provided through the environment variable ROBOCODE_SERVER_URL.
   *
   * @param botInfo is the bot info containing information about your bot.
   */
  public BaseBot(final BotInfo botInfo) {
    __baseBotInternals = new BaseBotInternals(this, botInfo, null);
  }

  /**
   * Constructor for initializing a new instance of the BaseBot class, which should be used
   * providing both the bot information and server URL for your bot.
   *
   * @param botInfo is the bot info containing information about your bot.
   * @param serverUrl is the server URL
   */
  public BaseBot(final BotInfo botInfo, URI serverUrl) {
    __baseBotInternals = new BaseBotInternals(this, botInfo, serverUrl);
  }

  /** {@inheritDoc} */
  @Override
  public final void start() {
    __baseBotInternals.start();
  }

  /** {@inheritDoc} */
  @Override
  public final void go() {
    __baseBotInternals.execute();
  }

  /** {@inheritDoc} */
  @Override
  public final String getVariant() {
    return __baseBotInternals.getVariant();
  }

  /** {@inheritDoc} */
  @Override
  public final String getVersion() {
    return __baseBotInternals.getVersion();
  }

  /** {@inheritDoc} */
  @Override
  public final int getMyId() {
    return __baseBotInternals.getMyId();
  }

  /** {@inheritDoc} */
  @Override
  public final String getGameType() {
    return __baseBotInternals.getGameSetup().getGameType();
  }

  /** {@inheritDoc} */
  @Override
  public final int getArenaWidth() {
    return __baseBotInternals.getGameSetup().getArenaWidth();
  }

  /** {@inheritDoc} */
  @Override
  public final int getArenaHeight() {
    return __baseBotInternals.getGameSetup().getArenaHeight();
  }

  /** {@inheritDoc} */
  @Override
  public final int getNumberOfRounds() {
    return __baseBotInternals.getGameSetup().getNumberOfRounds();
  }

  /** {@inheritDoc} */
  @Override
  public final double getGunCoolingRate() {
    return __baseBotInternals.getGameSetup().getGunCoolingRate();
  }

  /** {@inheritDoc} */
  @Override
  public final int getMaxInactivityTurns() {
    return __baseBotInternals.getGameSetup().getMaxInactivityTurns();
  }

  /** {@inheritDoc} */
  @Override
  public final int getTurnTimeout() {
    return __baseBotInternals.getGameSetup().getTurnTimeout();
  }

  /** {@inheritDoc} */
  @Override
  public final int getTimeLeft() {
    return __baseBotInternals.getTimeLeft();
  }

  /** {@inheritDoc} */
  @Override
  public final int getRoundNumber() {
    return __baseBotInternals.getCurrentTick().getRoundNumber();
  }

  /** {@inheritDoc} */
  @Override
  public final int getTurnNumber() {
    return __baseBotInternals.getCurrentTick().getTurnNumber();
  }

  /** {@inheritDoc} */
  @Override
  public final int getEnemyCount() {
    return __baseBotInternals.getCurrentTick().getEnemyCount();
  }

  /** {@inheritDoc} */
  @Override
  public final double getEnergy() {
    return __baseBotInternals.getCurrentTick().getBotState().getEnergy();
  }

  /** {@inheritDoc} */
  @Override
  public final boolean isDisabled() {
    return getEnergy() == 0;
  }

  /** {@inheritDoc} */
  @Override
  public final double getX() {
    return __baseBotInternals.getCurrentTick().getBotState().getX();
  }

  /** {@inheritDoc} */
  @Override
  public final double getY() {
    return __baseBotInternals.getCurrentTick().getBotState().getY();
  }

  /** {@inheritDoc} */
  @Override
  public final double getDirection() {
    return __baseBotInternals.getCurrentTick().getBotState().getDirection();
  }

  /** {@inheritDoc} */
  @Override
  public final double getGunDirection() {
    return __baseBotInternals.getCurrentTick().getBotState().getGunDirection();
  }

  /** {@inheritDoc} */
  @Override
  public final double getRadarDirection() {
    return __baseBotInternals.getCurrentTick().getBotState().getRadarDirection();
  }

  /** {@inheritDoc} */
  @Override
  public final double getSpeed() {
    return __baseBotInternals.getCurrentTick().getBotState().getSpeed();
  }

  /** {@inheritDoc} */
  @Override
  public final double getGunHeat() {
    return __baseBotInternals.getCurrentTick().getBotState().getGunHeat();
  }

  /** {@inheritDoc} */
  @Override
  public final Collection<BulletState> getBulletStates() {
    return __baseBotInternals.getCurrentTick().getBulletStates();
  }

  /** {@inheritDoc} */
  @Override
  public final Collection<? extends BotEvent> getEvents() {
    return __baseBotInternals.getCurrentTick().getEvents();
  }

  /** {@inheritDoc} */
  @Override
  public final void setTurnRate(double turnRate) {
    if (Double.isNaN(turnRate)) {
      throw new IllegalArgumentException("turnRate cannot be NaN");
    }
    __baseBotInternals.getBotIntent().setTurnRate(turnRate);
  }

  /** {@inheritDoc} */
  @Override
  public final double getTurnRate() {
    return __baseBotInternals.getCurrentTick().getBotState().getTurnRate();
  }

  /** {@inheritDoc} */
  @Override
  public final void setMaxTurnRate(double maxTurnRate) {
    __baseBotInternals.setMaxTurnRate(maxTurnRate);
  }

  /** {@inheritDoc} */
  @Override
  public final void setGunTurnRate(double gunTurnRate) {
    if (Double.isNaN(gunTurnRate)) {
      throw new IllegalArgumentException("gunTurnRate cannot be NaN");
    }
    __baseBotInternals.getBotIntent().setGunTurnRate(gunTurnRate);
  }

  /** {@inheritDoc} */
  @Override
  public final double getGunTurnRate() {
    return __baseBotInternals.getCurrentTick().getBotState().getGunTurnRate();
  }

  /** {@inheritDoc} */
  @Override
  public final void setMaxGunTurnRate(double maxGunTurnRate) {
    __baseBotInternals.setMaxGunTurnRate(maxGunTurnRate);
  }

  /** {@inheritDoc} */
  @Override
  public final void setRadarTurnRate(double radarTurnRate) {
    if (Double.isNaN(radarTurnRate)) {
      throw new IllegalArgumentException("radarTurnRate cannot be NaN");
    }
    __baseBotInternals.getBotIntent().setRadarTurnRate(radarTurnRate);
  }

  /** {@inheritDoc} */
  @Override
  public final double getRadarTurnRate() {
    return __baseBotInternals.getCurrentTick().getBotState().getRadarTurnRate();
  }

  /** {@inheritDoc} */
  @Override
  public final void setMaxRadarTurnRate(double maxRadarTurnRate) {
    __baseBotInternals.setMaxRadarTurnRate(maxRadarTurnRate);
  }

  /** {@inheritDoc} */
  @Override
  public final void setTargetSpeed(double targetSpeed) {
    if (Double.isNaN(targetSpeed)) {
      throw new IllegalArgumentException("targetSpeed cannot be NaN");
    }
    __baseBotInternals.getBotIntent().setTargetSpeed(targetSpeed);
  }

  /** {@inheritDoc} */
  @Override
  public final double getTargetSpeed() {
    return __baseBotInternals.getBotIntent().getTargetSpeed();
  }

  /** {@inheritDoc} */
  @Override
  public final void setMaxSpeed(double maxSpeed) {
    __baseBotInternals.setMaxSpeed(maxSpeed);
  }

  /** {@inheritDoc} */
  @Override
  public final boolean setFire(double firepower) {
    return __baseBotInternals.setFire(firepower);
  }

  /** {@inheritDoc} */
  @Override
  public final double getFirepower() {
    Double firepower = __baseBotInternals.getBotIntent().getFirepower();
    return firepower == null ? 0 : firepower;
  }

  /** {@inheritDoc} */
  @Override
  public final void setScan() {
    __baseBotInternals.getBotIntent().setScan(true);
  }

  /** {@inheritDoc} */
  @Override
  public final void setAdjustGunForBodyTurn(boolean adjust) {
    __baseBotInternals.getBotIntent().setAdjustGunForBodyTurn(adjust);
  }

  /** {@inheritDoc} */
  @Override
  public final boolean doAdjustGunForBodyTurn() {
    Boolean adjust = __baseBotInternals.getBotIntent().getAdjustGunForBodyTurn();
    return adjust != null && adjust;
  }

  /** {@inheritDoc} */
  @Override
  public final void setAdjustRadarForGunTurn(boolean adjust) {
    __baseBotInternals.getBotIntent().setAdjustRadarForGunTurn(adjust);
  }

  /** {@inheritDoc} */
  @Override
  public final boolean doAdjustRadarForGunTurn() {
    Boolean adjust = __baseBotInternals.getBotIntent().getAdjustRadarForGunTurn();
    return adjust != null && adjust;
  }

  /** {@inheritDoc} */
  @Override
  public final void addCustomEvent(Condition condition) {
    __baseBotInternals.addCondition(condition);
  }

  /** {@inheritDoc} */
  @Override
  public final void removeCustomEvent(Condition condition) {
    __baseBotInternals.removeCondition(condition);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getBodyColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getBodyColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setBodyColor(String bodyColor) {
    __baseBotInternals.getBotIntent().setBodyColor(bodyColor);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getTurretColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getTurretColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setTurretColor(String turretColor) {
    __baseBotInternals.getBotIntent().setTurretColor(turretColor);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getRadarColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getRadarColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setRadarColor(String radarColor) {
    __baseBotInternals.getBotIntent().setRadarColor(radarColor);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getBulletColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getBulletColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setBulletColor(String bulletColor) {
    __baseBotInternals.getBotIntent().setBulletColor(bulletColor);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getScanColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getScanColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setScanColor(String scanColor) {
    __baseBotInternals.getBotIntent().setScanColor(scanColor);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getTracksColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getTracksColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setTracksColor(String tracksColor) {
    __baseBotInternals.getBotIntent().setTracksColor(tracksColor);
  }

  /** {@inheritDoc} */
  @Override
  public final Integer getGunColor() {
    return __baseBotInternals.getCurrentTick().getBotState().getGunColor();
  }

  /** {@inheritDoc} */
  @Override
  public final void setGunColor(String gunColor) {
    __baseBotInternals.getBotIntent().setGunColor(gunColor);
  }

  /** {@inheritDoc} */
  @Override
  public final double calcMaxTurnRate(double speed) {
    return MAX_TURN_RATE - 0.75 * Math.abs(speed);
  }

  /** {@inheritDoc} */
  @Override
  public final double calcBulletSpeed(double firepower) {
    return 20 - 3 * firepower;
  }

  /** {@inheritDoc} */
  @Override
  public final double calcGunHeat(double firepower) {
    return 1 + (firepower / 5);
  }
}

File: robocode-tankroyale-bot-api-csharp/src/BaseBot.cs
using System;
using System.Collections.Generic;

namespace Robocode.TankRoyale.BotApi
{
  /// <summary>
  /// Abstract bot class that takes care of communication between the bot and the server, and sends
  /// notifications through the event handlers. Most bots can inherit from this class to get access
  /// to basic methods.
  /// </summary>
  public partial class BaseBot : IBaseBot
  {
    internal readonly BaseBotInternals __baseBotInternals;

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class, which should be used when
    /// both BotInfo and server URI is provided through environment variables, i.e., when starting
    /// up the bot using a bootstrap. These environment variables must be set to provide the server
    /// URL and bot information, and are automatically set by the bootstrap tool for Robocode.
    ///
    /// Example of how to set the predefined environment variables:
    ///
    /// ROBOCODE_SERVER_URI=ws://localhost<br/>
    /// BOT_NAME=MyBot<br/>
    /// BOT_VERSION=1.0<br/>
    /// BOT_AUTHOR=fnl<br/>
    /// BOT_DESCRIPTION=Sample bot<br/>
    /// BOT_URL=https://mybot.somewhere.net<br/>
    /// BOT_COUNTRY_CODE=DK<br/>
    /// BOT_GAME_TYPES=melee,1v1<br/>
    /// BOT_PROG_PLATFORM=.Net Core 3.1<br/>
    /// BOT_PROG_LANG=C# 8<br/>
    /// </summary>
    public BaseBot()
    {
      __baseBotInternals = new BaseBotInternals(this, null, null);
    }

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class, which should be used when
    /// server URI is provided through the environment variable ROBOCODE_SERVER_URL.
    /// </summary>
    /// <param name="botInfo">Is the bot info containing information about your bot.</param>
    public BaseBot(BotInfo botInfo)
    {
      __baseBotInternals = new BaseBotInternals(this, botInfo, null);
    }

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class, which should be used providing
    /// both the bot information and server URL for your bot.
    /// </summary>
    /// <param name="botInfo">Is the bot info containing information about your bot.</param>
    /// <param name="serverUrl">Is the server URI</param>
    public BaseBot(BotInfo botInfo, Uri serverUrl)
    {
      __baseBotInternals = new BaseBotInternals(this, botInfo, serverUrl);
    }

    /// <inheritdoc/>
    public void Start()
    {
      __baseBotInternals.Connect();
      __baseBotInternals.exitEvent.WaitOne();
    }

    /// <inheritdoc/>
    public void Go()
    {
      __baseBotInternals.SendIntent();
    }

    /// <inheritdoc/>
    public String Variant
    {
      get => __baseBotInternals.ServerHandshake.Variant;
    }

    /// <inheritdoc/>
    public String Version
    {
      get => __baseBotInternals.ServerHandshake.Version;
    }

    /// <inheritdoc/>
    public int MyId
    {
      get => __baseBotInternals.MyId;
    }

    /// <inheritdoc/>
    public String GameType
    {
      get => __baseBotInternals.GameSetup.GameType;
    }

    /// <inheritdoc/>
    public int ArenaWidth
    {
      get => __baseBotInternals.GameSetup.ArenaWidth;
    }

    /// <inheritdoc/>
    public int ArenaHeight
    {
      get => __baseBotInternals.GameSetup.ArenaHeight;
    }

    /// <inheritdoc/>
    public int NumberOfRounds
    {
      get => __baseBotInternals.GameSetup.NumberOfRounds;
    }

    /// <inheritdoc/>
    public double GunCoolingRate
    {
      get => __baseBotInternals.GameSetup.GunCoolingRate;
    }

    /// <inheritdoc/>
    public int? MaxInactivityTurns
    {
      get => __baseBotInternals.GameSetup.MaxInactivityTurns;
    }

    /// <inheritdoc/>
    public int TurnTimeout
    {
      get => __baseBotInternals.GameSetup.TurnTimeout;
    }

    /// <inheritdoc/>
    public int TimeLeft
    {
      get
      {
        long passesMicroSeconds = (DateTime.Now.Ticks - __baseBotInternals.TicksStart) / 10;
        return (int)(__baseBotInternals.GameSetup.TurnTimeout - passesMicroSeconds);
      }
    }

    /// <inheritdoc/>
    public int RoundNumber
    {
      get => __baseBotInternals.CurrentTick.RoundNumber;
    }

    /// <inheritdoc/>
    public int TurnNumber
    {
      get => __baseBotInternals.CurrentTick.TurnNumber;
    }

    /// <inheritdoc>
    public int EnemyCount
    {
      get => __baseBotInternals.CurrentTick.EnemyCount;
    }

    /// <inheritdoc/>
    public double Energy
    {
      get => __baseBotInternals.CurrentTick.BotState.Energy;
    }

    /// <inheritdoc/>
    public bool IsDisabled
    {
      get => Energy == 0;
    }

    /// <inheritdoc/>
    public double X
    {
      get => __baseBotInternals.CurrentTick.BotState.X;
    }

    /// <inheritdoc/>
    public double Y
    {
      get => __baseBotInternals.CurrentTick.BotState.Y;
    }

    /// <inheritdoc/>
    public double Direction
    {
      get => __baseBotInternals.CurrentTick.BotState.Direction;
    }

    /// <inheritdoc/>
    public double GunDirection
    {
      get => __baseBotInternals.CurrentTick.BotState.GunDirection;
    }

    /// <inheritdoc/>
    public double RadarDirection
    {
      get => __baseBotInternals.CurrentTick.BotState.RadarDirection;
    }

    /// <inheritdoc/>
    public double Speed
    {
      get => __baseBotInternals.CurrentTick.BotState.Speed;
    }

    /// <inheritdoc/>
    public double GunHeat
    {
      get => __baseBotInternals.CurrentTick.BotState.GunHeat;
    }

    /// <inheritdoc/>
    public IEnumerable<BulletState> BulletStates
    {
      get => __baseBotInternals.CurrentTick.BulletStates;
    }

    /// <inheritdoc/>
    public IEnumerable<Event> Events
    {
      get => __baseBotInternals.CurrentTick.Events;
    }

    /// <inheritdoc/>
    public double TurnRate
    {
      set
      {
        if (Double.IsNaN(value))
        {
          throw new ArgumentException("TurnRate cannot be NaN");
        }
        if (Math.Abs(value) > ((IBaseBot)this).MaxTurnRate)
        {
          value = ((IBaseBot)this).MaxTurnRate * Math.Sign(value);
        }
        __baseBotInternals.BotIntent.TurnRate = value;
      }
      get => __baseBotInternals.BotIntent.TurnRate ?? 0d;
    }

    /// <inheritdoc/>
    public double GunTurnRate
    {
      set
      {
        if (Double.IsNaN(value))
        {
          throw new ArgumentException("GunTurnRate cannot be NaN");
        }
        if (DoAdjustGunForBodyTurn)
        {
          value -= value;
        }
        if (Math.Abs(value) > ((IBaseBot)this).MaxGunTurnRate)
        {
          value = ((IBaseBot)this).MaxGunTurnRate * Math.Sign(value);
        }
        __baseBotInternals.BotIntent.GunTurnRate = value;
      }
      get => __baseBotInternals.BotIntent.GunTurnRate ?? 0d;
    }

    /// <inheritdoc/>
    public double RadarTurnRate
    {
      set
      {
        if (Double.IsNaN(value))
        {
          throw new ArgumentException("RadarTurnRate cannot be NaN");
        }
        if (DoAdjustRadarForGunTurn)
        {
          value -= value;
        }
        if (Math.Abs(value) > ((IBaseBot)this).MaxRadarTurnRate)
        {
          value = ((IBaseBot)this).MaxRadarTurnRate * Math.Sign(value);
        }
        __baseBotInternals.BotIntent.RadarTurnRate = value;
      }
      get => __baseBotInternals.BotIntent.RadarTurnRate ?? 0d;
    }

    /// <inheritdoc/>
    public double TargetSpeed
    {
      set
      {
        if (Double.IsNaN(value))
        {
          throw new ArgumentException("TargetSpeed cannot be NaN");
        }
        if (value > ((IBaseBot)this).MaxForwardSpeed)
        {
          value = ((IBaseBot)this).MaxForwardSpeed;
        }
        else if (value < ((IBaseBot)this).MaxBackwardSpeed)
        {
          value = ((IBaseBot)this).MaxBackwardSpeed;
        }
        __baseBotInternals.BotIntent.TargetSpeed = value;
      }
      get => __baseBotInternals.BotIntent.TargetSpeed ?? 0d;
    }

    /// <inheritdoc/>
    public bool SetFirepower(double firepower)
    {
      if (Double.IsNaN(firepower))
      {
        throw new ArgumentException("Firepower cannot be NaN");
      }
      if (GunHeat > 0)
      {
        return false; // cannot fire yet
      }
      if (firepower < ((IBaseBot)this).MinFirepower)
      {
        firepower = 0;
      }
      else if (firepower > ((IBaseBot)this).MaxFirepower)
      {
        firepower = ((IBaseBot)this).MaxFirepower;
      }
      __baseBotInternals.BotIntent.Firepower = firepower;
      return true;
    }

    /// <inheritdoc/>
    public bool DoAdjustGunForBodyTurn
    {
      set => __baseBotInternals.doAdjustGunForBodyTurn = value;
      get => __baseBotInternals.doAdjustGunForBodyTurn;
    }

    /// <inheritdoc/>
    public bool DoAdjustRadarForGunTurn
    {
      set => __baseBotInternals.doAdjustRadarForGunTurn = value;
      get => __baseBotInternals.doAdjustRadarForGunTurn;
    }

    /// <inheritdoc/>
    public int? GetBodyColor() => __baseBotInternals.CurrentTick.BotState.BodyColor;

    /// <inheritdoc/>
    public void SetBodyColor(string bodyColor) => __baseBotInternals.BotIntent.BodyColor = bodyColor;

    /// <inheritdoc/>
    public int? GetTurretColor() => __baseBotInternals.CurrentTick.BotState.TurretColor;

    /// <inheritdoc/>
    public void SetTurretColor(string turretColor) => __baseBotInternals.BotIntent.TurretColor = turretColor;

    /// <inheritdoc/>
    public int? GetRadarColor() => __baseBotInternals.CurrentTick.BotState.RadarColor;

    /// <inheritdoc/>
    public void SetRadarColor(string radarColor) => __baseBotInternals.BotIntent.RadarColor = radarColor;

    /// <inheritdoc/>
    public int? GetBulletColor() => __baseBotInternals.CurrentTick.BotState.BulletColor;

    /// <inheritdoc/>
    public void SetBulletColor(string bulletColor) => __baseBotInternals.BotIntent.BulletColor = bulletColor;

    /// <inheritdoc/>
    public int? GetScanColor() => __baseBotInternals.CurrentTick.BotState.ScanColor;

    /// <inheritdoc/>
    public void SetScanColor(string scanColor) => __baseBotInternals.BotIntent.ScanColor = scanColor;

    /// <inheritdoc/>
    public int? GetTracksColor() => __baseBotInternals.CurrentTick.BotState.TracksColor;

    /// <inheritdoc/>
    public void SetTracksColor(string tracksColor) => __baseBotInternals.BotIntent.TracksColor = tracksColor;

    /// <inheritdoc/>
    public int? GetGunColor() => __baseBotInternals.CurrentTick.BotState.GunColor;

    /// <inheritdoc/>
    public void SetGunColor(string gunColor) => __baseBotInternals.BotIntent.GunColor = gunColor;

    /// <inheritdoc/>
    public double CalcMaxTurnRate(double speed) => ((IBaseBot)this).MaxTurnRate - 0.75 * Math.Abs(speed);

    /// <inheritdoc/>
    public double CalcBulletSpeed(double firepower) => 20 - 3 * firepower;

    /// <inheritdoc/>
    public double CalcGunHeat(double firepower) => 1 + (firepower / 5);

    /// <inheritdoc/>
    public double CalcBearing(double direction) => NormalizeRelativeDegrees(direction - Direction);

    /// <inheritdoc/>
    public double CalcGunBearing(double direction) => NormalizeRelativeDegrees(direction - GunDirection);

    /// <inheritdoc/>
    public double CalcRadarBearing(double direction) => NormalizeRelativeDegrees(direction - RadarDirection);

    /// <inheritdoc/>
    public double NormalizeAbsoluteDegrees(double angle) => (angle %= 360) >= 0 ? angle : (angle + 360);

    /// <inheritdoc/>
    public double DirectionTo(double x, double y)
    {
      var dx = x - X;
      var dy = y - Y;
      return Math.Sqrt(dx * dx + dy * dy);
    }

    /// <inheritdoc/>
    public double DistanceTo(double x, double y) => NormalizeAbsoluteDegrees(Math.Atan2(x - X, y - Y));

    /// <inheritdoc/>
    public double NormalizeRelativeDegrees(double angle) => (angle %= 360) >= 0 ?
        ((angle < 180) ? angle : (angle - 360)) :
        ((angle >= -180) ? angle : (angle + 360));

    /// <inheritdoc/>
    public virtual void OnConnected(ConnectedEvent connectedEvent) => Console.WriteLine($"Connected to: {connectedEvent.ServerUri}");

    /// <inheritdoc/>
    public virtual void OnDisconnected(DisconnectedEvent disconnectedEvent) => Console.WriteLine($"Disconnected from: {disconnectedEvent.ServerUri}");

    /// <inheritdoc/>
    public virtual void OnConnectionError(ConnectionErrorEvent connectionErrorEvent) => Console.Error.WriteLine($"Connection error with: {connectionErrorEvent.ServerUri}");

    /// <inheritdoc/>
    public virtual void OnGameStarted(GameStartedEvent gameStatedEvent) { }

    /// <inheritdoc/>
    public virtual void OnGameEnded(GameEndedEvent gameEndedEvent) { }

    /// <inheritdoc/>
    public virtual void OnTick(TickEvent tickEvent) { }

    /// <inheritdoc/>
    public virtual void OnBotDeath(BotDeathEvent botDeathEvent) { }

    /// <inheritdoc/>
    public virtual void OnDeath(BotDeathEvent botDeathEvent) { }

    /// <inheritdoc/>
    public virtual void OnHitBot(BotHitBotEvent botHitBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnHitWall(BotHitWallEvent botHitWallEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletFired(BulletFiredEvent bulletFiredEvent) { }

    /// <inheritdoc/>
    public virtual void OnHitByBullet(BulletHitBotEvent bulletHitBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletHit(BulletHitBotEvent bulletHitBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletHitBullet(BulletHitBulletEvent bulletHitBulletEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletHitWall(BulletHitWallEvent bulletHitWallEvent) { }

    /// <inheritdoc/>
    public virtual void OnScannedBot(ScannedBotEvent scannedBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnSkippedTurn(SkippedTurnEvent skippedTurnEvent) { }

    /// <inheritdoc/>
    public virtual void OnWonRound(WonRoundEvent wonRoundEvent) { }

    /// <inheritdoc/>
    public virtual void OnCondition(Condition condition) { }
  }
}